week3/runoff.c

```diff
@@ -0,0 +1,210 @@
+            string name = get_string("Rank %i: ", candidate_number + 1);
+
+            if (!vote(voter_number, candidate_number, name))
+            {
+                printf("Invalid vote.\n");
+
+                return 4;
+            }
+        }
+
+        printf("\n");
+    }
+
+    while (true)
+    {
+        tabulate();
+
+        bool won = print_winner();
+        if (won)
+        {
+            break;
+        }
+
+        int min = find_min();
+        bool tie = is_tie(min);
+
+        if (tie)
+        {
+            for (int candidate_number = 0; candidate_number < candidate_count; candidate_number++)
+            {
+                if (!candidates[candidate_number].eliminated)
+                {
+                    printf("%s\n", candidates[candidate_number].name);
+                }
+            }
+
+            break;
+        }
+
+        eliminate(min);
+
+        for (int candidate_number = 0; candidate_number < candidate_count; candidate_number++)
+        {
+            candidates[candidate_number].votes = 0;
+        }
+    }
+
+    return 0;
+}
+
+bool vote(int voter, int rank, string name)
+{
+    for (int candidate_number = 0; candidate_number < candidate_count; candidate_number++)
+    {
+        if (strcmp(candidates[candidate_number].name, name) == 0)
+        {
+            preferences[voter][rank] = candidate_number;
+
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void tabulate(void)
+{
+    int first_vote;
+
+    for (int voter_number = 0; voter_number < voter_count; voter_number++)
+    {
+        for (int candidate_number = 0; candidate_number < candidate_count; candidate_number++)
+        {
+            first_vote = preferences[voter_number][candidate_number];
+
+            if (!candidates[first_vote].eliminated)
+            {
+                candidates[first_vote].votes++;
+                break;
+            }
+        }
+    }
+
+    return;
+}
+
+bool print_winner(void)
+{
+    for (int candidate_number = 0; candidate_number < candidate_count; candidate_number++)
+    {
+        if (candidates[candidate_number].votes > (voter_count/2))
+        {
+            printf("%s \n", candidates[candidate_number].name);
+
+            return true;
+        }
+    }
+
+    return false;
+}
+
+int find_min(void)
+{
+    int min_votes = voter_count;
+
+    for (int candidate_number = 0; candidate_number < candidate_count; candidate_number++)
+    {
+        if (candidates[candidate_number].votes < min_votes && !candidates[candidate_number].eliminated) 
+        {
+            min_votes = candidates[candidate_number].votes;
+        }
+    }
+
+    return min_votes;
+}
+
+bool is_tie(int min)
+{
+    for (int candidate_number = 0; candidate_number < candidate_count; candidate_number++)
+    {
+        if (!candidates[candidate_number].eliminated)
+        {
+            if (candidates[candidate_number].votes != min)
+            {
+                return false;
+            }
+        }
```

Comment:
```suggestion
        if (!candidates[candidate_number].eliminated && candidates[candidate_number].votes != min) return false;
```
Das selbe spielchen hier mit der Speaking Variable und dem verknüpften if statement.